controllers/keystoneapplicationcredential_controller.go

```diff
@@ -0,0 +1,401 @@
+	if err != nil {
+		return "", "", time.Time{}, fmt.Errorf("failed to create AC for user %s: %w", ac.Spec.UserName, err)
+	}
+
+	actualExpires := created.ExpiresAt
+	logger.Info("Created AC in Keystone",
+		"ACID", created.ID,
+		"userID", userID,
+		"expiresAt", actualExpires)
+	return created.ID, created.Secret, actualExpires, nil
+}
+
+func (r *ApplicationCredentialReconciler) storeACSecret(
+	ctx context.Context,
+	helperObj *helper.Helper,
+	ac *keystonev1.ApplicationCredential,
+	secretName, newID, newSecret string,
+) error {
+
+	data := map[string]string{
+		"AC_ID":     newID,
+		"AC_SECRET": newSecret,
+	}
+	tmpl := []util.Template{{
+		Name:       secretName,
+		Namespace:  ac.Namespace,
+		Type:       util.TemplateTypeNone,
+		CustomData: data,
+	}}
+	return oko_secret.EnsureSecrets(ctx, helperObj, ac, tmpl, nil)
+}
+
+// findUserIDAsAdmin uses admin-scoped token to look up user ID
+func (r *ApplicationCredentialReconciler) findUserIDAsAdmin(
+	ctx context.Context,
+	helperObj *helper.Helper,
+	keystoneAPI *keystonev1.KeystoneAPI,
+	userName string,
+) (string, error) {
+
+	logger := r.GetLogger(ctx)
+	adminOS, ctrlResult, err := keystonev1.GetAdminServiceClient(ctx, helperObj, keystoneAPI)
+	if err != nil {
+		return "", err
+	}
+	if ctrlResult != (ctrl.Result{}) {
+		return "", ErrAdminServiceClientNotReady
+	}
+
+	userObj, err := adminOS.GetUser(logger, userName, "Default")
+	if err != nil {
+		return "", fmt.Errorf("cannot find user %q: %w", userName, err)
+	}
+	return userObj.ID, nil
+}
+
+// shouldRotateNow returns true if we are already within GracePeriodDays
+func (r *ApplicationCredentialReconciler) shouldRotateNow(ac *keystonev1.ApplicationCredential) bool {
+	if ac.Status.ExpiresAt == nil || ac.Status.ExpiresAt.IsZero() {
+		return false
+	}
+
+	expiry := ac.Status.ExpiresAt.Time
+	grace := time.Duration(ac.Spec.GracePeriodDays) * 24 * time.Hour
+	rotateAt := expiry.Add(-grace)
+	return time.Now().After(rotateAt)
+}
+
+// computeNextRequeue schedules the next check to happen at "expiry - grace"
+func (r *ApplicationCredentialReconciler) computeNextRequeue(ac *keystonev1.ApplicationCredential) time.Duration {
+	// default requeue is 24h as minimal grace period is 1 day
+	defaultRequeue := 24 * time.Hour
+	if ac.Status.ExpiresAt == nil || ac.Status.ExpiresAt.IsZero() {
+		return defaultRequeue
+	}
+
+	expiry := ac.Status.ExpiresAt.Time
+
+	grace := time.Duration(ac.Spec.GracePeriodDays) * 24 * time.Hour
+	rotateAt := expiry.Add(-grace)
+
+	if rotateAt.Before(time.Now()) {
+		// Already within the grace window, so trigger rotation immediately
+		return 0
+	}
+
+	// Otherwise check again in 24 hours
+	return defaultRequeue
+}
+
+// randomSuffix generates a random string suffix
+func randomSuffix(n int) string {
+	b := make([]byte, (n+1)/2)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "fallback" // placeholder for generating failure
+	}
+	s := hex.EncodeToString(b)
+	return s[:n]
```

Comment:
Its unlikely, but should we check for collisions?

Reply:
should this function be a lib-common utility in the long term?